### PR TITLE
server: add vm boot details for start vm api

### LINKED ISSUE #5466: After installing 4.15.2.0 instances with uefi cannot boot for second time (xcp-ng 8.2)
##### ISSUE TYPE

 * Bug Report

##### COMPONENT NAME

~~~
CRITICAL - operation not possible
~~~

##### CLOUDSTACK VERSION
~~~
4.15.2.0
~~~

##### CONFIGURATION

Ubuntu server 18 with cloudstack 4.15.2.0
Hypervisor xcp-ng 8.2 with september security patch
Advanced networking

##### SUMMARY

Continuing from this issue
https://github.com/apache/cloudstack/issues/5204
Which is closed and this problem happening

After installation of 4.15.2.0, when creating a new instance there is an option in advanced setting for bios or uefi - legacy, secure (btw - why there is 2 options for uefi? on xen orchesra there are only bios and uefi)
When selecting bios everything is working perfect, even old vms from old templates which had issues with 4.15.1.0. You can shutdown, boot again, reboot, everything works as it should

When selectin uefi - both options (secure - legacy) the instance will boot without issue on the 1st boot. But if you shutdown the instance and try to boot it again there is issue. Without even touching something from settings. There is an error that 
"Unable to start a VM due to insufficient capacity"

image

and in the logs of cloudstack management there is something strange 

image

I have selected uefi and secure and haven't touched the settings. I can understand that it is trying to vm with bios boot option.

This happens even on old instances created with older versions of CS. Instance will work the 1st time booting but if you shut it down cannot boot anymore.

I think this is really critical.

##### STEPS TO REPRODUCE

~~~
1. Create new instance or boot an instance with option uefi.
2. shut it down.
3. Try to boot it and it won't boot anymore
~~~

##### EXPECTED RESULTS

~~~
The instance to boot as the first time.
I underline that i haven't touched anything in instance settings.
~~~

##### ACTUAL RESULTS

~~~
Instance cannot boot

~~~

### PR DESCRIPTION
**Pull request by shwstppr**
### Description

Fixes #5466

### Types of changes

- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] Enhancement (improves an existing feature and functionality)
- [ ] Cleanup (Code refactoring and cleanup, that may add test cases)

#### Feature/Enhancement Scale

- [ ] Major
- [ ] Minor

#### Bug Severity

- [ ] BLOCKER
- [ ] Critical
- [ ] Major
- [ ] Minor
- [ ] Trivial

### CODE REVIEW

**Comment by shwstppr** [comment]
@blueorangutan package

**Review by rhtyd — APPROVED** [review]
LGTM, needs testing

File: server/src/main/java/com/cloud/vm/UserVmManagerImpl.java [hunk 1]
```diff
@@ -3005,15 +3004,30 @@ protected boolean applyUserData(HypervisorType hyperVisorType, UserVm vm, Nic ni
...
+                s_logger.trace(String.format("Adding boot options (%s, %s, %s) into the param map for VM start as UEFI detail(%s=%s) found for the VM",
+                        VirtualMachineProfile.Param.UefiFlag.getName(),
+                        VirtualMachineProfile.Param.BootType.getName(),
+                        VirtualMachineProfile.Param.BootMode.getName(),
+                        uefiDetail.getName(),
+                        uefiDetail.getValue()));
+            }
+            additonalParams.put(VirtualMachineProfile.Param.UefiFlag, "Yes");
+            additonalParams.put(VirtualMachineProfile.Param.BootType, uefiDetail.getName());
+            additonalParams.put(VirtualMachineProfile.Param.BootMode, uefiDetail.getValue());
```
Comment (shwstppr) [inline server/src/main/java/com/cloud/vm/UserVmManagerImpl.java:3029]: Similar to startvm method call during `deployVirtualMachine` https://github.com/apache/cloudstack/blob/main/server/src/main/java/com/cloud/vm/UserVmManagerImpl.java#L4599-L4601

**Comment by shwstppr** [comment]
@blueorangutan test matrix

**Comment by weizhouapache** [comment]
@shwstppr target to 4.15 ?

**Comment by shwstppr** [comment]
Done @weizhouapache 

@blueorangutan package

**Comment by shwstppr** [comment]
@weizhouapache done

@blueorangutan package

File: server/src/main/java/com/cloud/vm/UserVmManagerImpl.java [hunk 2]
```diff
@@ -2884,15 +2889,30 @@ protected boolean applyUserData(HypervisorType hyperVisorType, UserVm vm, Nic ni
...
+            additonalParams = new HashMap<>();
             if (s_logger.isTraceEnabled()) {
                 s_logger.trace(String.format("Adding %s into the param map", VirtualMachineProfile.Param.BootIntoSetup.getName()));
             }
 
             additonalParams.put(VirtualMachineProfile.Param.BootIntoSetup, cmd.getBootIntoSetup());
         }
+        UserVmDetailVO uefiDetail = userVmDetailsDao.findDetail(cmd.getId(), ApiConstants.BootType.UEFI.toString());
+        if (uefiDetail != null) {
+            if (additonalParams == null) {
```
Comment (weizhouapache) [inline server/src/main/java/com/cloud/vm/UserVmManagerImpl.java:2901]: would it be better to do ?

```
Map<VirtualMachineProfile.Param, Object> additonalParams =  new HashMap<>();
```

File: server/src/main/java/com/cloud/vm/UserVmManagerImpl.java [hunk 3]
```diff
@@ -2884,15 +2889,30 @@ protected boolean applyUserData(HypervisorType hyperVisorType, UserVm vm, Nic ni
...
+            }
+            if (s_logger.isTraceEnabled()) {
+                s_logger.trace(String.format("Adding boot options (%s, %s, %s) into the param map for VM start as UEFI detail(%s=%s) found for the VM",
+                        VirtualMachineProfile.Param.UefiFlag.getName(),
+                        VirtualMachineProfile.Param.BootType.getName(),
+                        VirtualMachineProfile.Param.BootMode.getName(),
+                        uefiDetail.getName(),
+                        uefiDetail.getValue()));
+            }
+            additonalParams.put(VirtualMachineProfile.Param.UefiFlag, "Yes");
```
Comment (weizhouapache) [inline server/src/main/java/com/cloud/vm/UserVmManagerImpl.java:2912]: @shwstppr 
can you extract this to method ? that's Daan often asked :-D

File: server/src/main/java/com/cloud/vm/UserVmManagerImpl.java [hunk 4]
```diff
@@ -16,6 +16,7 @@
 // under the License.
 package com.cloud.vm;
 
+import static com.cloud.configuration.ConfigurationManagerImpl.VM_USERDATA_MAX_LENGTH;
```
Comment (weizhouapache) [inline server/src/main/java/com/cloud/vm/UserVmManagerImpl.java:19]: these import are not needed

Comment (shwstppr) [reply server/src/main/java/com/cloud/vm/UserVmManagerImpl.java:19]: could be something during rebase, removing

**Comment by shwstppr** [comment]
@weizhouapache can you please have a look again

**Comment by weizhouapache** [comment]
@blueorangutan package

**Comment by shwstppr** [comment]
@blueorangutan test matrix

**Comment by weizhouapache** [comment]
@blueorangutan test centos7 xcpng82 keepEnv

**Comment by shwstppr** [comment]
Not sure why tests getting following exception:
```
2021-09-21 13:08:19,019 - CRITICAL - EXCEPTION: test_deploy_vm_volume_creation_failure: ['Traceback (most recent call last):\n', '  File "/usr/lib64/python2.7/unittest/case.py", line 360, in run\n    self.setUp()\n', '  File "/marvin/tests/smoke/misc/test_deploy_vm.py", line 75, in setUp\n    count=6)\n', '  File "/usr/lib/python2.7/site-packages/marvin/lib/base.py", line 5267, in create\n    cmd = configureSimulator.configureSimulatorCmd()\n', "NameError: global name 'configureSimulator' is not defined\n"]
```
Will be investigating

**Comment by weizhouapache** [comment]
> test_deploy_vm_start_failure

@shwstppr this should has been fixed by #5456, maybe trillian testing env does not have the latest integration test scripts.

File: server/src/main/java/com/cloud/vm/UserVmManagerImpl.java [hunk 5]
```diff
@@ -2882,17 +2896,18 @@ protected boolean applyUserData(HypervisorType hyperVisorType, UserVm vm, Nic ni
     @Override
     @ActionEvent(eventType = EventTypes.EVENT_VM_START, eventDescription = "starting Vm", async = true)
     public UserVm startVirtualMachine(StartVMCmd cmd) throws ExecutionException, ConcurrentOperationException, ResourceUnavailableException, InsufficientCapacityException, ResourceAllocationException {
-        Map<VirtualMachineProfile.Param, Object> additonalParams = null;
+        Map<VirtualMachineProfile.Param, Object> additonalParams = new HashMap<>();
         if (cmd.getBootIntoSetup() != null) {
-            if (additonalParams == null) {
-                additonalParams = new HashMap<>();
-            }
+            additonalParams = new HashMap<>();
```
Comment (weizhouapache) [inline server/src/main/java/com/cloud/vm/UserVmManagerImpl.java:2901]: 2901 can be removed

**Review by weizhouapache — APPROVED** [review]
code lgtm

left a comment

**Comment by shwstppr** [comment]
@blueorangutan package

**Comment by rhtyd** [comment]
@shwstppr does this need testing? Pl kick appropriate env for smoketests + manual tests. Thanks.

**Comment by weizhouapache** [comment]
> @shwstppr does this need testing? Pl kick appropriate env for smoketests + manual tests. Thanks.

@rhtyd 
the previous trillian test failures are caused by broken tests. I think smoketests is not needed.
of course we can also re-build the packages and re-kick the test when @DaanHoogland has backported #5456 to 4.15.
let's wait for a while.

**Comment by rhtyd** [comment]
@blueorangutan package

**Comment by nvazquez** [comment]
@blueorangutan test matrix

**Review by nvazquez — APPROVED** [review]
Code LGTM, needs testing

**Comment by sureshanaparti** [comment]
@blueorangutan test centos7 xenserver-71

**Comment by weizhouapache** [comment]
@blueorangutan test centos7 xcpng82 keepEnv

**Comment by rhtyd** [comment]
ping @shwstppr @weizhouapache @nvazquez - is this ready for merging?
Looks like the usage test failure is intermittent?

**Comment by weizhouapache** [comment]
> ping @shwstppr @weizhouapache @nvazquez - is this ready for merging?
> Looks like the usage test failure is intermittent?

@rhtyd 
it is fine with me.

**Comment by shwstppr** [comment]
@shwstppr ready from my side

**Comment by shwstppr** [comment]
@rhtyd ready from my side

**Pull request merged**